Add DMD (standard and exact)

Jonas-Nicodemus opened the pull request:
### New Feature Dynamic Mode Decomposition (DMD)

This PR replaces #1267 and adds a basic variant of the DMD algorithm (cf. TRLBK14, PBK14) and corresponding demos.

Note that we removed the random DMD as discussed.

pmli commented:
Hi @Jonas-Nicodemus, thanks for the PR.

I think it would be good to first clean up the commits a bit. Some of them add things that are later removed (e.g., randomized DMD and functions for amplitudes and Vandermonde matrix) and it would be good to have commit messages consistent with the current style in the `main` branch (i.e., "[location] description"). This could be done using `git rebase`, but probably the easiest way is with `git reset`:

```bash
git checkout feature-dmd
git reset upstream/main
git add [file]; git commit -m "[location] description"; ...
git push -f origin feature-dmd
```

TiKeil commented:
> Hi @Jonas-Nicodemus, thanks for the PR.
> 
> I think it would be good to first clean up the commits a bit. Some of them add things that are later removed (e.g., randomized DMD and functions for amplitudes and Vandermonde matrix) and it would be good to have commit messages consistent with the current style in the `main` branch (i.e., "[location] description"). This could be done using `git rebase`, but probably the easiest way is with `git reset`:
> 
> ```shell
> git checkout feature-dmd
> git reset upstream/main
> git add [file]; git commit -m "[location] description"; ...
> git push -f origin feature-dmd
> ```

I am not sure whether `git reset` is a good way here. It removes the fact that the original work has been done by Henrike.
However, I agree that it would be some work to remove commits that add features that are deleted afterwards.

pmli commented on src/pymor/algorithms/dmd.py:
```diff
@@ -0,0 +1,98 @@
+    omega = np.log(evals) / dt
+
+    # ordering
+    sort_idx = np.argsort(np.abs(omega))[::-1]
```
The sorting here is done using `omega`, which corresponds to the eigenvalues of the continuous-time system. So, sorting by `omega.real` (or `np.abs(evals)`) would make more sense, right?

sdrave replied:
I'd agree ..

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,89 @@
+
+    # ----- Testing DMD -----
+    W1, E1 = dmd(X=U, modes='standard')
+    W2, E2 = dmd(X=U, svd_rank=5, modes='exact')
```
I found that `W2` is complex (the GUI raised a warning about converting complex numbers to real). It would be good to check why this happens.

sdrave replied:
@pmli, maybe I'm confused, but why should `W2` (or `W1`) not be complex? The builtin visualizers don't support complex numbers, so it would be better to explicitly take the real part or so.

pmli replied:
I just found it weird that `W2` is complex and `W1` is real (I thought they should be of the same `dtype`).

sdrave replied:
Are you sure? They both are complex on my machine ..

pmli replied:
Hm, I don't know what happened, but now both are complex for me...
Anyway, there is still the issue of plotting complex-valued solutions.

sdrave replied:
Probably it would make sense to plot the real and imaginary parts side by side ..

sdrave replied:
Done.

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,89 @@
+    W2, E2 = dmd(X=U, svd_rank=5, modes='exact')
+
+    print('Visualize ...')
+    m.visualize(W1, title='DMD Modes - standard')
+    plt.plot(E1.real, E1.imag, 'b.')
+    plt.title('DMD Eigenvalues - standard')
```
```suggestion
    m.visualize(W1, title='DMD Modes (Standard)')
    plt.plot(E1.real, E1.imag, '.')
    plt.title('DMD Eigenvalues (Standard)')
```

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,89 @@
+    plt.ylabel('Imaginary')
+    plt.show()
+
+    m.visualize(W2, title='5 DMD Modes - exact')
+    plt.plot(E2.real, E2.imag, 'b.')
+    plt.title('DMD Eigenvalues - exact')
```
```suggestion
    m.visualize(W2, title='Five DMD Modes (Exact)')
    plt.plot(E2.real, E2.imag, '.')
    plt.title('DMD Eigenvalues (Exact)')
```

sdrave commented on src/pymor/algorithms/dmd.py:
```diff
@@ -0,0 +1,98 @@
+        assert False
+
+    if return_A_approx:
+        A_approx = LowRankOperator(U, A_tilde, U)
```
I guess, both `A_tilde` and `A_approx` can be interesting for different reason. So why not have separate options for returning `A_tilde` and `A_approx`?

sdrave commented:
I'd say this is now almost ready to go. Only one more thing: I think the we don't really need the `burgers_dmd` and `parabolic_dmd` demo, which do basically the main thing. So I would suggest to get rid of one of these or have one demo where both models can be selected. Regarding the models, probably both aren't optimal for showcasing DMD, but I don't have a better suggestion ATM. Do you @pmli? @Jonas-Nicodemus, we are planning to have the next release one week before Christmas. Do you think you can spare some time to take care of the remaining minor issues?

Jonas-Nicodemus commented:
> I'd say this is now almost ready to go. Only one more thing: I think the we don't really need the `burgers_dmd` and `parabolic_dmd` demo, which do basically the main thing. So I would suggest to get rid of one of these or have one demo where both models can be selected. Regarding the models, probably both aren't optimal for showcasing DMD, but I don't have a better suggestion ATM. Do you @pmli? @Jonas-Nicodemus, we are planning to have the next release one week before Christmas. Do you think you can spare some time to take care of the remaining minor issues?

Sure, I will take a look at the remaining issues this week.

pmli commented:
> Regarding the models, probably both aren't optimal for showcasing DMD, but I don't have a better suggestion ATM. Do you @pmli?

No, not really. I agree about removing one of the demos.

sdrave commented:
@Jonas-Nicodemus, @pmli I did some further polishing of the DMD code. Notable changes:

- I fixed a bug in the computation of the exact DMD modes.
- `svd_rank` -> `modes` + `atol`, `rtol`
- `dt` -> `cont_time_dt`, only take logarithm when argument is specified
- allow to choose between ordering by magnitude or frequency
- I somewhat extended the burgers demo and switched to 1d where it is easier to see what is going on.

Please take a look if you agree with the changes. From my perspective, this is now ready to be merged.

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,90 @@
+    tic = time.perf_counter()
+    U = m.solve(mu)
+    print(f'Solving took {time.perf_counter() - tic}s')
+    m.visualize(U)
```
Adding a `title` would be good.

sdrave replied:
Done.

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,90 @@
+        c = plt.Circle((0., 0.), 1., linestyle='--', color='k', fill=False)
+        xl = max(np.max(np.abs(E1.real)), 1.) * 1.1
+        yl = max(np.max(np.abs(E1.imag)), 1.) * 1.1
+        plt.gca().add_artist(c)
+        plt.xlim([-xl, xl])
+        plt.ylim([-yl, yl])
```
```suggestion
        plt.gca().add_artist(c)
        plt.axis('square')
        plt.xlim([-xl, xl])
        plt.ylim([-yl, yl])
```

pmli commented on src/pymordemos/burgers_dmd.py:
```diff
@@ -0,0 +1,90 @@
+
+    print('Visualize ...')
+    # plot DMD modes
+    m.visualize((W1.real, W1.imag, W2.real, W2.imag),
+                legend=('standard - real part', 'standard - imaginary part',
+                        'exact - real part', 'exact - imaginary part'),
+                title='DMD Modes')
```
It seems that the modes are always overlapping. Is there a way to make the exact modes dashed?

sdrave replied:
No, the line style cannot be customized in the visualizer.

pmli replied:
That's unfortunate. I guess that separating it into two visualizations isn't so helpful, and I don't have other ideas to improve it.

sdrave commented on src/pymor/algorithms/dmd.py:
```diff
@@ -0,0 +1,122 @@
+    A_approx
+        |LowRankOperator| containing the approximation of the operator `A` with `AX=Y`.
+    A_tilde
+         Low-rank dynamics.
```
Done.

sdrave commented on src/pymor/algorithms/dmd.py:
```diff
@@ -0,0 +1,122 @@
+            `Wk = U * evecs`, where `U` are the left singular vectors of `X`.
+        - 'exact' : computes the exact dynamic modes, `Wk = (1/evals) * Y * V * Sigma_inv * evecs`.
+    order
+        Sort DMD eigenvalues either by `'magnitude'` or `'frequency'`.
```
Done

sdrave commented on src/pymor/algorithms/dmd.py:
```diff
@@ -0,0 +1,122 @@
+    assert isinstance(X, VectorArray)
+    assert isinstance(Y, VectorArray) or Y is None
+    assert modes is None or modes <= X.dim
+    assert modes is None or modes < len(X)
```
Right. Actually, I decided to remove the assertion completely and similar assertions for the svd/pod algs: Even when you specify `modes` you need to be aware that you might get less modes due to atol/rtol. So I think it is overall better to never fail. (The user should see the argument as an upper bound.)

Jonas-Nicodemus commented:
> @Jonas-Nicodemus, @pmli I did some further polishing of the DMD code. Notable changes:
> 
> * I fixed a bug in the computation of the exact DMD modes.
> * `svd_rank` -> `modes` + `atol`, `rtol`
> * `dt` -> `cont_time_dt`, only take logarithm when argument is specified
> * allow to choose between ordering by magnitude or frequency
> * I somewhat extended the burgers demo and switched to 1d where it is easier to see what is going on.
> 
> Please take a look if you agree with the changes. From my perspective, this is now ready to be merged.

Good morning @sdrave,
I just took a look and totally agree with the changes. The burgers demo now makes more sense, and thanks for taking care of the additional review from @pmli. Let me know if there is anything left.
